java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -60,308 +63,333 @@
+            result.add(executionState.push(operatorResult));
+          } else {
+            // condition is unknown. split with constraints.
+            result.add(new ExecutionState(executionState).setRelation(leftValue, operator.negate(), rightValue).push(SymbolicValue.BOOLEAN_FALSE));
+            result.add(new ExecutionState(executionState).setRelation(leftValue, operator, rightValue).push(SymbolicValue.BOOLEAN_TRUE));
+          }
         }
       }
-      currentResult.unknownStates.add(currentState);
     }
 
     @Override
-    public final void visitInstanceOf(InstanceOfTree tree) {
-      currentResult.unknownStates.add(currentState);
+    public final void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+      super.visitArrayAccessExpression(tree);
+      for (ExecutionState executionState : currentStates) {
+        executionState.pop();
+        executionState.pop();
+        executionState.push(createSymbolicInstanceValue());
+      }
     }
 
     @Override
-    public final void visitLiteral(LiteralTree tree) {
-      if ("false".equals(tree.value())) {
-        currentResult.falseStates.add(currentState);
-      } else if ("true".equals(tree.value())) {
-        currentResult.trueStates.add(currentState);
-      } else {
-        currentResult.unknownStates.add(currentState);
+    public final void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      super.visitAssignmentExpression(tree);
+      for (ExecutionState executionState : currentStates) {
+        SymbolicValue expression = executionState.pop();
+        executionState.pop();
+        Symbol.VariableSymbol symbol = extractVariableSymbol(tree.variable());
+        if (symbol != null) {
+          executionState.assignValue(symbol, expression);
+        }
+        executionState.push(expression);
       }
     }
 
     @Override
-    public final void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
-      if (isSuperOrThisMemberSelect(tree)) {
-        scan(tree.identifier());
+    public final void visitBinaryExpression(BinaryExpressionTree tree) {
+      if (tree.is(Tree.Kind.CONDITIONAL_AND)) {
+        evaluateConditionalAnd(tree);
+      } else if (tree.is(Tree.Kind.CONDITIONAL_OR)) {
+        evaluateConditionalOr(tree);
+      } else if (tree.is(Tree.Kind.EQUAL_TO)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.EQUAL_TO);
+      } else if (tree.is(Tree.Kind.GREATER_THAN)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.GREATER_THAN);
+      } else if (tree.is(Tree.Kind.GREATER_THAN_OR_EQUAL_TO)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.GREATER_EQUAL);
+      } else if (tree.is(Tree.Kind.LESS_THAN)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.LESS_THAN);
+      } else if (tree.is(Tree.Kind.LESS_THAN_OR_EQUAL_TO)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.LESS_EQUAL);
+      } else if (tree.is(Tree.Kind.NOT_EQUAL_TO)) {
```

Comment:
@merciesa if values are dense, then switch on enum will be faster than sequence of if-else-if